presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchClient.java

```diff
@@ -0,0 +1,264 @@
+        requireNonNull(schemaName, "schemaName is null");
+        requireNonNull(tableName, "tableName is null");
+        try {
+            for (Map.Entry<SchemaTableName, ElasticsearchTableDescription> entry : tableDescriptions.entrySet()) {
+                buildColumns(entry.getKey(), entry.getValue());
+            }
+        }
+        catch (Exception e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, e.getMessage());
+        }
+        return tableDescriptions.get(new SchemaTableName(schemaName, tableName));
+    }
+
+    public ClusterSearchShardsResponse getSearchShards(ElasticsearchTableDescription tableDescription)
+    {
+        Client client = clients.get(tableDescription.getClusterName());
+        ClusterSearchShardsRequest request = new ClusterSearchShardsRequest(tableDescription.getIndex());
+        try {
+            return client.admin().cluster().searchShards(request).get();
+        }
+        catch (InterruptedException | ExecutionException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, e.getMessage());
+        }
+    }
+
+    private void buildColumns(SchemaTableName schemaTableName, ElasticsearchTableDescription tableDescription)
+            throws ExecutionException, InterruptedException, IOException, JSONException
+    {
+        Set<ElasticsearchColumn> columns = new HashSet();
+
+        Client client = clients.get(tableDescription.getClusterName());
+        GetMappingsRequest mappingsRequest = new GetMappingsRequest().types(tableDescription.getType());
+        String index = tableDescription.getIndex();
+
+        if (index != null && !index.isEmpty()) {
+            mappingsRequest.indices(index);
+        }
+        ImmutableOpenMap<String, ImmutableOpenMap<String, MappingMetaData>> mappings = client.admin().indices().getMappings(mappingsRequest).get().getMappings();
+
+        Iterator<String> indexIterator = mappings.keysIt();
+        while (indexIterator.hasNext()) {
+            MappingMetaData mappingMetaData = mappings.get(indexIterator.next()).get(tableDescription.getType());
+            JSONObject mappingObject = new JSONObject(mappingMetaData.source().toString()).getJSONObject(tableDescription.getType());
+            JSONObject properties = mappingObject.getJSONObject("properties");
+            JSONObject meta = mappingObject.getJSONObject("_meta");
+            JSONArray arrays = meta.getJSONArray("lists");
+            List<String> lists = new ArrayList<String>();
+            for (int i = 0; i < arrays.length(); i++) {
+                lists.add(arrays.getString(i));
+            }
+
+            for (String columnMetadata : getColumnsMetadata(null, properties)) {
+                ElasticsearchColumn column = createColumn(columnMetadata, lists);
+                List<String> columnNames = columns.stream().map(ElasticsearchColumn::getName).collect(Collectors.toList());
+                if (column != null && !columnNames.contains(column.getName())) {
```

Comment:
you are doing two passes: one with the `columns.stream()` and the other one with `columnNames.contains()`. You can do this with a single pass.

Reply:
will fix